cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
 	SkipPhases []string
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// UpgradeConfiguration contains a list of elements that is specific "kubeadm upgrade"-only runtime information.
+type UpgradeConfiguration struct {
+	metav1.TypeMeta
+
+	// ApiServerManifestPath is the path to API server manifest
+	// Default: /etc/kubernetes/manifests/kube-apiserver.yaml
+	ApiServerManifestPath string
+
+	// AllowExperimentalUpgrades instruct kubeadm to show unstable versions of Kubernetes as an upgrade
+	// alternative and allow upgrading to an alpha/beta/release candidate version of Kubernetes.
+	AllowExperimentalUpgrades bool
+
+	// AllowRCUpgrades instruct kubeadm to show release candidate versions of Kubernetes as an upgrade alternative and
+	// allow upgrading to a release candidate version of Kubernetes.
+	AllowRCUpgrades bool
+
+	// CertificateRenewal instructs kubeadm to execute certificate renewal during upgrades
+	// Default: true
+	CertificateRenewal *bool
+
+	// ClusterConfiguration holds the cluster-wide information, and embeds that struct (which can be (un)marshalled separately as well)
+	ClusterConfiguration `json:"-"`
```

Comment:
> is it too complex to just pass upgrade configuration separately from cluster configuration around separately instead of this embedding?

I will take a look and get back to you later on this.